src/hotspot/share/opto/addnode.cpp

```diff
@@ -1004,30 +991,88 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
   const TypeInt *r0 = t0->is_int(); // Handy access
   const TypeInt *r1 = t1->is_int();
 
-  // Complementing a boolean?
-  if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if (r0->is_con() && r1->is_con()) {
+    // Constant fold: (c1 ^ c2) -> c3
+    return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  }
 
-  if( !r0->is_con() || !r1->is_con() ) // Not constants
-    return TypeInt::INT;        // Any integer, but still no symbols.
+  // At least one of the arguments is not constant
+
+  // Result of xor can only have bits sets where any of the
+  // inputs have bits set. lo can always become 0.
+
+  if (r0->_lo >= 0 && r1->_lo >= 0) {
+      jint max = calc_xor_max(r0->_hi, r1->_hi);
+      return TypeInt::make(0, max, MAX2(r0->_widen, r1->_widen));
+  }
 
-  // Otherwise just XOR them bits.
-  return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  return TypeInt::INT;        // Any integer, but still no symbols.
 }
 
+jint XorINode::calc_xor_max(const jint hi_0, const jint hi_1) {
```

Comment:
Since this is only used in `XorINode::Value`, I think a static local method would be better. And if you are comfortable with templates, I suggest templatizing it to unify the implementation with `jlong` below.

Reply:
I've templatized it and took it out of the class, but didn't make it static, because I need the test to call it. Can I do better than that?